ocaml/libs/http-lib/server_io.ml

```diff
@@ -26,15 +26,83 @@ type handler = {
   ; lock: Semaphore.Counting.t
 }
 
+module WorkerPool = struct
+  type task = unit -> unit
+
+  type t = {
+      queue: task Queue.t
+    ; size: int
+    ; lock: Mutex.t
+    ; cond_non_empty: Condition.t
+  }
+
+  let worker p =
+    let next () =
+      Mutex.lock p.lock ;
+      while Queue.is_empty p.queue do
+        Condition.wait p.cond_non_empty p.lock
+      done ;
+      let task = Queue.pop p.queue in
+      Mutex.unlock p.lock ;
+      ignore (task ())
+    in
+    while true do
+      next ()
+    done
+
+  let create ~size =
+    let queue = Queue.create () in
+    let lock = Mutex.create () in
+    let cond_non_empty = Condition.create () in
+    let pool = {queue; size; lock; cond_non_empty} in
+    ignore (List.init size (fun _ -> Thread.create worker pool)) ;
+    pool
+
+  let enqueue p task =
+    if Mutex.try_lock p.lock then (
+      let len = Queue.length p.queue in
+      if len = p.size then (
```

Comment:
Are you exactly sure is what you really want?
The tasks are queued into `p.queue` and removed ASAP from a worker thread if possible.
So potentially you reach `p.size` when you have `p.size` tasks running and `p.size` tasks pending, so `2 * p.size`.